### PR TITLE
Membrane channel refactor

### PR DESCRIPTION
**Pull request by bknueven**
## Summary/Motivation:
Separate out various calculations in RO to "Membrane Channel" control volume.

## Changes proposed in this PR:
-
-

### Legal Acknowledgement

By contributing to this software project, I agree to the following terms and conditions for my contribution:

1. I agree my contributions are submitted under the license terms described in the LICENSE.txt file at the top level of this directory.
2. I represent I am authorized to make the contributions and grant the license. If my employer has rights to intellectual property that includes these contributions, I represent that I have received permission to make contributions and grant the required license on behalf of that employer.

### CODE REVIEW

File: watertap/core/membrane_channel0d.py
Comment (TimBartholomew) [inline watertap/core/membrane_channel0d.py]: "has_length_and_width" is probably better aligned with naming conventions in IDAES

File: watertap/core/membrane_channel0d.py
Comment (TimBartholomew) [inline watertap/core/membrane_channel0d.py]: I'm assuming this is for co-current flow and counter-current flow, however, I'm wary of this phrasing. Flow direction always goes from inlet to outlet, its just for counter-current flow the inlet and outlet are on opposite sides.

I'll see how this is implemented, but I don't think we want a state block or port named with inlet - actually being the outlet for the counter flow.

File: watertap/core/membrane_channel1d.py
Comment (TimBartholomew) [inline watertap/core/membrane_channel1d.py:57]: Is this a new configuration argument? I can see the utility of it, but I don't know if we have an example using it.

File: watertap/core/membrane_channel1d.py
Comment (TimBartholomew) [inline watertap/core/membrane_channel1d.py]: This a minor comment - but we probably want to lower this default. Usually theres not much benefit going beyond 10, well honestly even 5. This figure comes from one of my papers for the error we can expect from typical operation of membrane processes compared to 100 nodes - it also includes a monte carlo across multiple parameters to define the distribution: 
image

File: watertap/core/membrane_channel1d.py
Comment (TimBartholomew) [inline watertap/core/membrane_channel1d.py:138]: Ok I like this better. Could we do a similar set up for the 0D model?

Replace flow_information config argument with this. I guess we would still have the issue with the control volume properties_in and properties_out having the wrong names, but the reference would be fine. I'm not sure whats the best approach.

File: watertap/core/membrane_channel_base.py
Comment (TimBartholomew) [inline watertap/core/membrane_channel_base.py]: We might want to change this wording. We aren't necessarily doing a volumetric flowrate balance, we are just providing extensive flow information to the interface. Maybe "add_extensive_flow_to_interface" would be best.

File: watertap/core/membrane_channel_base.py
Comment (TimBartholomew) [inline watertap/core/membrane_channel_base.py]: Ohh interesting, so the concentration polarization equation will live in different places depending on the configuration arguments. I see why that's necessary for initialization, I'm wondering if we always just want the concentration polarization equation on the unit for it to not move based on config arguments.

File: watertap/unit_models/reverse_osmosis_base.py
Comment (TimBartholomew) [inline watertap/unit_models/reverse_osmosis_base.py:157]: This is all for backwards compatibility right? I think that's a good approach for this PR, but in the near future we may want to change our tests to the new format and not support this. At least for WaterTAP we don't have many external users who would be mad about this breaking change.

File: watertap/unit_models/tests/test_reverse_osmosis_1D.py
Comment (TimBartholomew) [inline watertap/unit_models/tests/test_reverse_osmosis_1D.py]: Yeah we can probably just get rid of these types of tests. If an expected object doesn't exist or isn't the right type, eventually we would get a pyomo error in the build.

**Review by TimBartholomew — COMMENTED** [review]
Looks good to me. The only unexpected consequence for me with making the flux be a unit model variable is that the concentration polarization equation then has to be on the unit model. I think that's fine though.

**Comment by TimBartholomew** [comment]
Oh and the other thing being the trouble with setting up the naming for counterflow operation in the 0D.

I think since the CV0D automatically creates properties_in and properties_out, maybe the best way to handle this is if it is counterflow the reference properties (to make it quack like 1D) could change whether properties_in or properties_out is assigned to 0 or 1. I'm not sure if that is your current implementation.

File: watertap/core/membrane_channel0d.py
Comment (andrewlee94) [reply watertap/core/membrane_channel0d.py]: This look like it is referring to a feature in the core IDAES code, which allows for systems where it is more robust to solve a process in reverse order (e.g. some refrigeration cycles). It is not related (directly) to co- or counter-current flow.

I am not sure whether this feature is required here, and unless there is a need for it I would suggest it might be best to drop it.

File: watertap/core/membrane_channel1d.py
Comment (andrewlee94) [reply watertap/core/membrane_channel1d.py:57]: This is also a core IDAES feature - this makes me wonder if this should be importing a standard config block from somewhere in IDAES, as it looks like it is re-creating its own.

Also, do not feel that you need to duplicate all config argument from IDAES - there are far more than you will need for any one case, and you should only implement those that you need. The underlying framework will default to the most general case if no input is given.

File: watertap/core/membrane_channel_base.py
Comment (andrewlee94) [reply watertap/core/membrane_channel_base.py]: I would second this - having things move around makes code maintenance and testing much harder. If you can put the polarization equation in the same place across all models, it will make your life much easier in future.

Comment (bknueven) [reply watertap/core/membrane_channel_base.py]: So, we could move the flux variable, but *not* the flux calculation, to the membrane channel. That would allow the concentration polarization equation to move to the membrane channel as well.

Since the flux calculation depends on the state on the "oppose side", (e.g., in the current RO units the permeate stateblock, in future OARO on another membrane channel), it necessarily has to stay on the unit model.

This gets at the heart of the "seam" question. I think adding a flux variable to the channel is a good idea, but of course it's up to the modeller to calculate it in a reasonable fashion.

@adam-a-a what do you think?

File: watertap/unit_models/reverse_osmosis_base.py
Comment (bknueven) [reply watertap/unit_models/reverse_osmosis_base.py:157]: Yes. I agree we should probably break it, and should be in a separate PR as it requires touching a huge number of files.

File: watertap/unit_models/tests/test_reverse_osmosis_1D.py
Comment (bknueven) [reply watertap/unit_models/tests/test_reverse_osmosis_1D.py]: I agree, done in 68bd79d.

File: watertap/core/membrane_channel0d.py
Comment (bknueven) [reply watertap/core/membrane_channel0d.py]: Done in [8e9eda1](https://github.com/watertap-org/watertap/pull/679/commits/8e9eda1c8aad94d68ac8bb2e616968fdd7b2c530).

File: watertap/core/membrane_channel1d.py
Comment (bknueven) [reply watertap/core/membrane_channel1d.py]: Done in [75001f7](https://github.com/watertap-org/watertap/pull/679/commits/75001f7d0e210eb40b208911be74cfb8b55990c8)

File: watertap/core/membrane_channel_base.py
Comment (bknueven) [reply watertap/core/membrane_channel_base.py]: Done in [4300d89](https://github.com/watertap-org/watertap/pull/679/commits/4300d89a88728121b1e43b98bbcfedf862b6b384)

File: watertap/core/membrane_channel0d.py
Comment (bknueven) [reply watertap/core/membrane_channel0d.py]: Right, the only reason it is here is for the call to `super().add_state_blocks` -- I gave the overloaded methods the same signature as that in the control volume. However, if we think for membrane channels we'll always use the default, I'm happy to eliminate the option.

File: watertap/core/membrane_channel1d.py
Comment (bknueven) [reply watertap/core/membrane_channel1d.py:57]: I think it's a hold-over from 1DRO (which this is copied from) which wasn't used. ~~Will eliminate.~~ Removed in [115392b](https://github.com/watertap-org/watertap/pull/679/commits/115392b850bddde937b06d259997cdfe2519fdc4).

Comment (adam-a-a) [reply watertap/core/membrane_channel1d.py:57]: I just want to note that this was added to RO1D since the area can technically vary across the spatial domain. Although `DistributedVars.variant` wasn't being used, I included this for a potential future application, if we ever want to consider variable area across the length of the membrane channel.

File: watertap/core/membrane_channel_base.py
Comment (adam-a-a) [inline watertap/core/membrane_channel_base.py:394]: Should we get rid of this if we are adding the `eq_concentration_polarization` constraint when CP type is none?

File: watertap/core/membrane_channel_base.py
Comment (adam-a-a) [inline watertap/core/membrane_channel_base.py:360]: should there be an elif above this, basically checking that CP type is not none? If not none, then create cp_modulus

File: watertap/core/membrane_channel_base.py
Comment (adam-a-a) [reply watertap/core/membrane_channel_base.py]: Well, I am thinking we'll want to have this on the unit model and that it would be better to have `eq_concentration_polarization` in the same place across config types. 

Moreover, `eq_concentration_polarization`, at a minimum, will differ based on the flow direction and whether the channel is facing the active side of the membrane or the porous support layer of the membrane, which I would think is decided at the unit model level. While it'd be nice to have the CP constraint for each channel control volume, I think it'd be ok to let this one slide and have it setup at the unit model level.

File: watertap/core/membrane_channel_base.py
Comment (bknueven) [reply watertap/core/membrane_channel_base.py:394]: I agree that these lines are not strictly necessary.

It's just some defensive programming against somebody adding a fourth CP type and forgetting to update this method.

Or worse, passing in `"foo"` as the CP type.

Comment (adam-a-a) [reply watertap/core/membrane_channel_base.py:394]: Nevermind- you're returning in the bit before this so this only happens when someone specifies some alternative option besides the 3 that we've defined.

File: watertap/core/membrane_channel_base.py
Comment (bknueven) [reply watertap/core/membrane_channel_base.py]: One thing to keep in mind is the membrane channel could have different build options for these different configurations (e.g., which side of the membrane am I on, which direction is the flow going).

Comment (adam-a-a) [reply watertap/core/membrane_channel_base.py]: OK, so running with that, how about handling a custom relationship for `eq_concentration_polarization`?

EDIT: Would it make a difference to the developer whether this would need to be applied in the membrane channel or unit model?

Comment (bknueven) [reply watertap/core/membrane_channel_base.py]: Who is doing the customizing?

If we put the flux variables in membrane channel, you would add another option there, I think.

Comment (adam-a-a) [reply watertap/core/membrane_channel_base.py]: I was just imagining if someone wanted to use a custom relationship to account for CP, e.g., some empirical relationship related to some particular module geometry. In any case, I suppose we could have an arg for custom_CP_con or something like that for the add_concentration_polarization method.

Comment (adam-a-a) [reply watertap/core/membrane_channel_base.py]: So maybe we should see how things work out if we put the flux vars on the channel, even though they would technically be shared between adjacent channels.

Comment (bknueven) [reply watertap/core/membrane_channel_base.py]: > EDIT: Would it make a difference to the developer whether this would need to be applied in the membrane channel or unit model?


Ah, for adding a new one, no.

But, for utilizing an existing one in a new unit model, yes. Instead of copying+pasting, you call the appropriate method (with the appropriate options) on the membrane channel.

This is the crux of what @andrewlee94 is saying above.

File: watertap/core/membrane_channel1d.py
Comment (bknueven) [reply watertap/core/membrane_channel1d.py:138]: Done in 68d8ae1, which also makes the membrane channel, and by extension the RO unit model, more robust to the transformation scheme (forward basically wasn't supported before).

File: watertap/unit_models/reverse_osmosis_0D.py
Comment (adam-a-a) [inline watertap/unit_models/reverse_osmosis_0D.py]:
```suggestion
    def _add_feed_side_membrane_channel_and_geometry(self):
```

File: watertap/unit_models/reverse_osmosis_1D.py
Comment (adam-a-a) [inline watertap/unit_models/reverse_osmosis_1D.py]:
```suggestion
    def _add_feed_side_membrane_channel_and_geometry(self):
```

File: watertap/unit_models/reverse_osmosis_1D.py
Comment (adam-a-a) [inline watertap/unit_models/reverse_osmosis_1D.py]: I don't think there was any harm in leaving the area definition since a variable area could come into play in the future.

File: watertap/core/membrane_channel_base.py
Comment (adam-a-a) [inline watertap/core/membrane_channel_base.py]: Should we delete this comment or did you want to keep as an example?

File: watertap/examples/flowsheets/lsrro/lsrro.py
Comment (adam-a-a) [inline watertap/examples/flowsheets/lsrro/lsrro.py:766]: Why did we need to add these `raise_on_failure=False` args? Did something affect LSRRO initialization? Or was this for debugging purposes?

File: watertap/unit_models/reverse_osmosis_base.py
Comment (adam-a-a) [inline watertap/unit_models/reverse_osmosis_base.py]:
```suggestion
        self._add_feed_side_membrane_channel_and_geometry()
```

File: watertap/core/membrane_channel0d.py
Comment (TimBartholomew) [reply watertap/core/membrane_channel0d.py]: Oh its not related to co- or counter-current flow? I think it can be eliminated then, although I don't know anything about the benefits of solving a process in reverse order (e.g. Andrew's comment about refrigeration cycles).

File: watertap/core/membrane_channel0d.py
Comment (TimBartholomew) [inline watertap/core/membrane_channel0d.py:277]: Minor comment and it shouldn't hold up the PR, but maybe we want to include a warning that the area scaling factor is being set by default. dp_dx probably doesn't need this warning because its an intensive variable and this is probably a good scaling factor for all practical uses of RO.

File: watertap/core/membrane_channel1d.py
Comment (TimBartholomew) [reply watertap/core/membrane_channel1d.py:138]: Wait, I may have misinterpreted this. In my original review I thought this flow_direction was actually the direction of flow and could be used by us to set co-current or counter-current for OARO (i.e. the permeate channel would have backward flow). But Andrew's comment suggest that this is just the direction of the solve rather than the actual flow. I'm not sure, but we can figure this all out with implementing the OARO model.

File: watertap/core/membrane_channel0d.py
Comment (andrewlee94) [reply watertap/core/membrane_channel0d.py]: This feature has never actually been used in IDAES - it was added based on a suggestion/request from a team member for future applications.

File: watertap/core/membrane_channel1d.py
Comment (TimBartholomew) [inline watertap/core/membrane_channel1d.py]: At first I thought the difference here was because of the naming conventions of the 1D model. But I'm confused, should for the 0D and 1D model this is the pressure loss per unit length? Shouldn't they have the same scaling factor?

File: watertap/examples/flowsheets/RO_with_energy_recovery/RO_with_energy_recovery.py
Comment (TimBartholomew) [inline watertap/examples/flowsheets/RO_with_energy_recovery/RO_with_energy_recovery.py:277]: Were we previously initializing with 1 degree of freedom? Oops :)

File: watertap/unit_models/reverse_osmosis_base.py
Comment (TimBartholomew) [inline watertap/unit_models/reverse_osmosis_base.py:441]: Ok so this concentration polarization constraint is added on the unit model and the constraint for the other concentration polarization types (i.e. fixed, none) are on the membrane channel? I don't know if there is a better solution.

What's the downside with moving the constraints for the other concentration polarization types to the unit model? I understand that the channel initialization won't have concentration polarization constraint when it could have for fixed and none options, but for the calculated option the channel initialization we would have to assume the interface is at the same state as the bulk - which is also fine for the fixed and none options.

**Review by TimBartholomew — COMMENTED** [review]
A few comments, I think we just need a quick discussion on the location of the concentration polarization constraints but otherwise I think this is good.

File: watertap/core/membrane_channel1d.py
Comment (andrewlee94) [reply watertap/core/membrane_channel1d.py:138]: @TimBartholomew Note that there are two similarly named argument - `flow_direction` and `information_flow`. `flow_direction` is use for the physical flow (basically does the `d/dx` term need a -1 in front of it).

File: watertap/unit_models/reverse_osmosis_base.py
Comment (bknueven) [reply watertap/unit_models/reverse_osmosis_base.py:441]: @adam-a-a discussed this and thought this was the best, though not great, solution.

Putting the flux on the channel doesn't make much sense as it is the mass transfer term. This constraint is also different on different sides of the channel. However, all the other concentration polarization equations are the same (as far as I understand) on both sides of the channel, so conceptually this separation makes some sense. In any case, we now create a `cp_modulus` Var in every case except `none`, so in some ways the channel doesn't really care beyond "create cp_modulus or not" -- and this constraint serves to calculate it.

We could put them all the unit model; however, note that any equations we put on the unit model need to be re-implemented in each unit.

File: watertap/unit_models/tests/test_reverse_osmosis_0D.py
Comment (adam-a-a) [inline watertap/unit_models/tests/test_reverse_osmosis_0D.py:267]: Should the test also be checking for `eq_concentration_polarization` since that is built at unit model level in ro base, only when CP type is calculated? Also, I am guessing assertions on the cp_modulus constraint would be for the future membrane channel tests

File: watertap/examples/flowsheets/RO_with_energy_recovery/RO_with_energy_recovery.py
Comment (bknueven) [reply watertap/examples/flowsheets/RO_with_energy_recovery/RO_with_energy_recovery.py:277]: This snuck is as part of #690, so it hasn't been there long ...

File: watertap/core/membrane_channel1d.py
Comment (bknueven) [reply watertap/core/membrane_channel1d.py]: We should probably think through this -- it's just copied from the analogous code in the current 1DRO model: 

https://github.com/watertap-org/watertap/blob/2d2a8fdca20acc0a343433afa511396fd788a91f/watertap/unit_models/reverse_osmosis_1D.py#L561-L566

File: watertap/unit_models/reverse_osmosis_base.py
Comment (andrewlee94) [reply watertap/unit_models/reverse_osmosis_base.py:441]: Does this suggest we should have a common utility function/method for creating these constraints at the unit level then (e.g. on a common base class for the models)?

**Review by adam-a-a — COMMENTED** [review]
I'm glad the tests work out thus far and I think we'll be able to start OARO soon. I will say though that the code is a bit trickier to navigate now, just in terms of knowing where things live and the order in which they connect. Hopefully, it will just be a matter of getting used to the changes.

File: watertap/unit_models/reverse_osmosis_base.py
Comment (adam-a-a) [reply watertap/unit_models/reverse_osmosis_base.py:441]: @TimBartholomew I think we now only have this `eq_concentration_polarization` at the unit level when CP type is calculated. When none or fixed, we now only have a cp_modulus constraint in the membrane channel CV which uses the cp_modulus to get interface concentration. When CP type is none, concentrations at bulk and interface are set equal using `eq_cp_modulus`.

Comment (adam-a-a) [reply watertap/unit_models/reverse_osmosis_base.py:441]: I think what we have now is fine actually. In the membrane channel CV, we just use a cp_modulus constraint, which makes sense from a channel perspective (i.e., each channel will have its own cp_modulus w/respect to length across the channel, and this is just the ratio of interface to bulk concentration). 

However, the constraint for calculating the membrane interface concentration can take on different forms, depending on whether we are using a first principles relationship, an empirical relationship, or depending on _which_ channel and flow configuration we are considering. Thus, it makes sense for any detailed calculation for CP to be at the unit model; meanwhile, we do no such calculation at the channel and merely compute interface concentration by an assumed ratio. 

Overall, we no longer have the same constraint living in either the unit model or channel model when switching between Enum types for concentration polarization. `eq_concentration_polarization` only lives on the unit model and `eq_cp_modulus` only lives on the channel CV.

File: watertap/examples/flowsheets/RO_with_energy_recovery/RO_with_energy_recovery.py
Comment (adam-a-a) [reply watertap/examples/flowsheets/RO_with_energy_recovery/RO_with_energy_recovery.py:277]: Yeah, there probably should be tests for the RO/ERD flowsheet to prevent that in the future.

File: watertap/unit_models/reverse_osmosis_base.py
Comment (bknueven) [reply watertap/unit_models/reverse_osmosis_base.py]: Always catching our spelling mistakes :-)

File: watertap/examples/flowsheets/lsrro/lsrro.py
Comment (bknueven) [reply watertap/examples/flowsheets/lsrro/lsrro.py:766]: This was essentially the default behavior of the RO model before, which this PR changes to be more IDAES compliant. Note the `fail_on_warning` flag:
https://github.com/watertap-org/watertap/blob/2d2a8fdca20acc0a343433afa511396fd788a91f/watertap/unit_models/reverse_osmosis_0D.py#L353-L362
https://github.com/watertap-org/watertap/blob/2d2a8fdca20acc0a343433afa511396fd788a91f/watertap/unit_models/reverse_osmosis_1D.py#L407-L416

In any case, for LSRRO sometimes these units fail to initialize on the first or second pass, especially at high salinities where the first-stage probably *is* infeasible. So nothing has really changed besides the default RO initialization flag, which I think is more correct in this PR.

File: watertap/examples/flowsheets/RO_with_energy_recovery/RO_with_energy_recovery.py
Comment (bknueven) [reply watertap/examples/flowsheets/RO_with_energy_recovery/RO_with_energy_recovery.py:277]: The RO initialization will now fail if the degrees of freedom aren't 0, which it wasn't doing before. (Which is how I found this.)

File: watertap/core/membrane_channel0d.py
Comment (bknueven) [reply watertap/core/membrane_channel0d.py]: I've remove it as part of 495fbfa9

File: watertap/core/membrane_channel1d.py
Comment (adam-a-a) [reply watertap/core/membrane_channel1d.py]: So I traced this back to PR #162, back in July of last year. I have no memory of why I did this, but if I had to guess, it was likely out of desperation when toggling between `has_pressure_drop` True or False and running into some scaling issue at the time. It was my 4th month so probably an unnecessary inclusion.

File: watertap/unit_models/reverse_osmosis_1D.py
Comment (bknueven) [reply watertap/unit_models/reverse_osmosis_1D.py]: I've reverted the removal of this in 495fbfa.

File: watertap/core/membrane_channel1d.py
Comment (bknueven) [reply watertap/core/membrane_channel1d.py:57]: @adam-a-a would like to maintain this, so I reverted the change that removed it. I don't have strong feelings either way.

File: watertap/core/membrane_channel1d.py
Comment (bknueven) [reply watertap/core/membrane_channel1d.py]: You are correct -- `dP_dx` will only be created if there's a pressure change.

In the event that you do not, I'm guessing these `pressure_dx` variables will be small, hence the scaling factor. I'm inclined to leave it as-is with an additional comment. Thoughts?

File: watertap/core/membrane_channel_base.py
Comment (bknueven) [reply watertap/core/membrane_channel_base.py]: I updated the comment, which got find-replace mangled.

File: watertap/unit_models/tests/test_reverse_osmosis_0D.py
Comment (bknueven) [reply watertap/unit_models/tests/test_reverse_osmosis_0D.py:267]: I think @TimBartholomew and I had a conversation about removing tests for specific components broadly, and relying more on performance values.

I'm happy to add these back in for those components that still live on the RO model, though I often find they're a bit more trouble than they're worth -- especially those checking specific constraint names, which don't particularly affect the API of the model.

File: watertap/core/membrane_channel0d.py
Comment (bknueven) [reply watertap/core/membrane_channel0d.py:277]: Interestingly we don't do this currently -- note that this is the cross-sectional area in this context (to match the name with 1DCV).

**Review by TimBartholomew — APPROVED** [review]

**Comment by bknueven** [comment]
> I will say though that the code is a bit trickier to navigate now, just in terms of knowing where things live and the order in which they connect. Hopefully, it will just be a matter of getting used to the changes.

FWIW I agree -- the benefit will come in (1) more rapid development of unit models and (2) consolidating our other existing membrane models to extend and use the membrane channel CV.

File: watertap/core/membrane_channel1d.py
Comment (adam-a-a) [reply watertap/core/membrane_channel1d.py]: An additional comment would be helpful if we leave it in. However, I am wondering if we should even have to do this or not. That is, is `feed_side.pressure_dx` unintentionally being created somewhere by us (e.g., faulty if statement logic), or is this just how things are setup in ControlVolume1D? If the latter, I think it's fine to keep; if the former, maybe we should provide a fix.

**Comment by bknueven** [comment]
Good point — I will look into it

File: watertap/core/membrane_channel1d.py
Comment (bknueven) [reply watertap/core/membrane_channel1d.py]: Removed in https://github.com/watertap-org/watertap/pull/679/commits/a86e13148826e885990157b680b3ac6a22b9eaa2.

Based on the way the 1DCV is set-up, I think we still need these variables to enforce the isobaric assumption when there's no pressure change. Or rather, we could get rid of them but we would need to enforce that constraint ourselves -- does not seem worth the trouble.

**Review by adam-a-a — APPROVED** [review]
I think we can merge this and if anything needs to be resolved, we can address issues as OARO implementation is pursued. Furthermore, we should eventually include tests for membrane_channel classes.

**Pull request merged**
